### DISCUSSION TITLE
LoadNanostring()/ReadNanostring() Updates

### DISCUSSION THREAD

**Discussion by swbioinf** [satijalab/seurat#9261] [General]
Hi,

Is anyone actively working on LoadNanostring()/ReadNanostring() functions atm? 

I'm considering on working on some changes/improvements/bugfixes on these functions to address the following.

* IMPROVEMENT: Add pass-through of metadata feilds from loadNanostring to ReadNanostring (see https://github.com/satijalab/seurat/issues/8895) 
* (EXISTING PR FIX): Removing 'set list' of metadata feilds - not all are present and other columns are (e.g. a la cart probes). I can see @roanvanscheppingen  has already fixed this in an outstanding pull request https://github.com/satijalab/seurat/pull/8900  .
* IMPROVEMENT: Maybe add 'fov' as a default metadata feild, it seems a bit odd that this isn't read as default
* BUFIX: Read and maintain counts matrix as a sparse matrix - I'm getting OOMs on a moderately sized datatset (needs 144Gb RAM to load successfully, that's ~so~ 2018! :) ).

Other minor suggestions welcome.

This should address some of the differences between the initial 'example data' released by nanostring, and what is now being generated by production cosmx machines. 

Currently I'm just working off a hacked up local version of these functions. 

If its not actively being worked on I'll try to build a pull request - but don't want to duplicate effort!

Cheers.
Sarah.

**Comment by roanvanscheppingen** [comment]
Thanks for the tag. I've not been pushing further fixes to these functions, but rather have my own hacked version as well, or do workarounds where I build my own Seurats from Nanostring flatfiles. 

The PR is still valid though. Feel free to expand on this PR or use the fix in your own PR. Keep in mind that Nanostring tends to swap around their annotation quite a lot still. https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/posts/flat-file-exports/flat-files-compare.html

Best, 
Roan

**Reply by swbioinf** [reply]
Thanks @roanvanscheppingen - yes I wonder how many of us have these local hacks :) Working as part of a facility though I'm keen to get the default working! I might adopt and build on your PR then. Thanks!

I certainly wouldn't complain if nanostring updated their format to mtx though!

**Reply by roanvanscheppingen** [reply]
Hi, 

While doing the great cleanup I was a bit too enthusiastic and deleted my fork of the seurat repo, thus closing my PR request. Sorry for the inconvenience...

**Comment by swbioinf** [comment]
Tagging this to say I am working on it, though its not on github yet.

**Reply by AlbertoFabbri93** [reply]
Do you have any news on this?

**Reply by swbioinf** [reply]
Late response, but it all got sidelined when I got caught up in some analysis. Same old story!

Don't quite remember where I was at when I put it down, but have just used it on another dataset, and its seemingly functional, though not quite ready for a pull request. 

Its certainly not fully tested.

I've attached the current version - functions called 'ReadNanostring.X' and 'LoadNanostring.X' : 
[LoadNanostring_R.txt](https://github.com/user-attachments/files/21310699/LoadNanostring_R.txt)



* It pulls in all metadata fields by default. This fixes the lack of fov also.
* It ignores the count matrix file and gets cell counts from the molecules file
* Uses fread to read molecules...
* And added a tempdir parameter to avoid this fread issue when the local /tmp is too small: https://www.linkedin.com/pulse/trivial-fix-after-3-hours-debugging-kirill-tsyganov/
* The ReadNanostring  fov.filter parameter is ignored
* Some old benchmarking suggests its faster and uses less ram, but I think this might be a little optimistic, and my notes suggest its quite variable. Take with largeish grain of salt.

```

| Test                                                 | elapsed Min | Peak_RAM_Used_G |
|------------------------------------------------------|-------------|-----------------|
| New ReadNanostring - smaller file                    | 3           | 7.7             |
| Existing ReadNanostring -   smaller file             | 4           | 18.7            |
| New ReadNanostring - larger   file                   | 13          | 14.8            |
| Existing ReadNanostring -   larger slide             | 18          | 67.4            |
| Existing   LoadNanostring - old small dataset        | 1           | 1.9             |
| Standard LoadNanostring - old small   dataset        | 1           | 2.0             |
```

I'll see how this goes with the latest dataset, and try to move toward something contributeable. But if anyone wants to adopt this into a pull request please do.
